doc/source/raysgd/raysgd_pytorch.rst

```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
 
+  1. If any worker node is lost, Ray will mark the training task as complete (``ray.wait`` will return).
+  2. Ray will throw ``RayActorException`` when fetching the result for any worker, so the Trainer class will call ``ray.get`` on the "finished" training task.
+  3. Upon catching this exception, the Trainer class will kill all of its workers.
+  4. The Trainer will then detect the quantity of available resources (either CPUs or GPUs). It will then restart as many workers as it can, each resuming from the last checkpoint. Note that this may result in fewer workers than initially specified.
+  5. If there are no available resources, the Trainer will apply an exponential backoff before retrying to create workers.
+  6. If there are available resources and the Trainer has fewer workers than initially specified, then it will scale up its worker pool until it reaches the initially specified ``num_workers``.
 
-Hyperparameter Optimization on Distributed Pytorch
---------------------------------------------------
+Note that we assume the Trainer itself is not on a pre-emptible node. It is currently not possible to recover from a Trainer node failure.
+
+Users can set ``checkpoint="auto"`` to always checkpoint the current model before executing a pass over the training dataset.
+
+.. code-block:: bash
+
+    trainer.train(max_retries=N, checkpoint="auto")
+
+
+Advanced: Hyperparameter Tuning
+-------------------------------
 
 ``PyTorchTrainer`` naturally integrates with Tune via the ``PyTorchTrainable`` interface. The same arguments to ``PyTorchTrainer`` should be passed into the ``tune.run(config=...)`` as shown below.
 
-.. literalinclude:: ../../../python/ray/experimental/sgd/examples/tune_example.py
+.. literalinclude:: ../../../python/ray/experimental/sgd/pytorch/examples/tune_example.py
    :language: python
    :start-after: __torch_tune_example__
 
 
-Package Reference
------------------
+Simultaneous Multi-model training
+---------------------------------
+
+In certain scenarios such as training GANs, you may want to use multiple models in the training loop. You can do this in the ``PyTorchTrainer`` by allowing the ``model_creator`` and the ``optimizer_creator`` to return multiple values. You can see the `DCGAN script <https://github.com/ray-project/ray/blob/master/python/ray/experimental/sgd/pytorch/examples/dcgan.py>`_ for an end-to-end example.
+
+.. code-block:: python
 
-.. autoclass:: ray.experimental.sgd.pytorch.PyTorchTrainer
-    :members:
+    def model_creator(config):
+        netD = Discriminator()
+        netD.apply(weights_init)
 
-    .. automethod:: __init__
+        netG = Generator()
+        netG.apply(weights_init)
+        return netD, netG
```

Comment:
I wonder if we should return lists when more than 1 is returned? Or should it always return a list for uniformity, even when just 1 is returned? Similarly, should any function that accepts model(s) as an argument always take a list?

Reply:
Well, I guess in python the return value is always a list/iterable if there's more than 1 returned. 

I think the general case will be to return 1 item, so forcing a list return is quite cumbersome. Internally, it is converted to a list though.

Reply:
```
In [1]: def test():
   ...:     return 1, 2, 3
   ...:

In [2]: a = test()

In [3]: a
Out[3]: (1, 2, 3)

Reply:
To be clear this is a valid point - we can open a separate issue/get user feedback down the road.